docs(admission-controller): add chart migration howto (#808)
kubewarden/docs: docs/admission-controller/version-1.37/modules/en/pages/howtos/chart-migration.adoc

include::partial$variables.adoc[]

= Migration from three-chart setup
:page-languages: [en, de, es, fr, ja, pt, zh]
:revdate: 2026-06-26
:page-revdate: {revdate}
:description: Migrate from the legacy kubewarden-crds, kubewarden-controller, and kubewarden-defaults Helm charts to the unified admission-controller chart.
:doc-persona: ["kubewarden-operator"]
:doc-topic: ["chart-migration"]
:doc-type: ["howto"]
:keywords: ["{admc-name}", "kubewarden", "helm", "chart", "migration", "upgrade", "kubewarden-crds", "kubewarden-controller", "kubewarden-defaults", "admission-controller"]
:sidebar_label: Migration from three-chart setup
:sidebar_position: 10
:current-version: {page-origin-branch}

If you are running the legacy three-chart setup (`kubewarden-crds`,
`kubewarden-controller`, `kubewarden-defaults`), you have two ways to move to the
single chart. The migration script
(`kubewarden-unified-adm-controller-chart-migration.sh`) does it without
admission downtime. Otherwise you can uninstall the old charts and reinstall the

> new one, which stops policy evaluation while you do it.

Reviewer: ```suggestion
Starting from Kubewarden 1.37, a single helm chart is used to manage
the stack.

If you are running the legacy three-chart setup (`kubewarden-crds`,
`kubewarden-controller`, `kubewarden-defaults`), you have upgrade paths:

* Use the migration script (`kubewarden-unified-adm-controller-chart-migration.sh`):
  this performs the migration without any downtime.
* Follow the reinstallation instructions: cleaner approach, but involves a short
  time during which your cluster won't be protected by policies.
```

== Manual reinstallation

This is the simplest path, but policies stop being evaluated while it runs. You
back up the policies and policy servers, uninstall the old {admc-short-name} stack,

> install the new chart, then reapply the backups.

Reviewer: ```suggestion
This is the simplest approach. It consists of:

1. Doing a backup of your policies
2. Uninstall the old helm charts
3. Install the new single helm chart
4. Restore all your policies

During step 2 and 4 your cluster won't be protected by any policy.
```

+
[source,sh]
----
FILTER='del(.items[].metadata.uid, .items[].metadata.resourceVersion, .items[].metadata.creationTimestamp, .items[].metadata.generation, .items[].metadata.managedFields, .items[].status)'

kubectl get clusteradmissionpolicies -A -o yaml | yq "$FILTER" > clusteradmissionpolicies-backup.yaml
kubectl get admissionpolicies -A -o yaml | yq "$FILTER" > admissionpolicies-backup.yaml
kubectl get clusteradmissionpolicygroups -A -o yaml | yq "$FILTER" > clusteradmissionpolicygroups-backup.yaml
kubectl get admissionpolicygroups -A -o yaml | yq "$FILTER" > admissionpolicygroups-backup.yaml
kubectl get policyservers -A -o yaml | yq "$FILTER" > policyservers-backup.yaml
----

. Uninstall the old Helm charts:
+
[source,sh]
----
helm uninstall kubewarden-defaults -n kubewarden
helm uninstall kubewarden-controller -n kubewarden
helm uninstall kubewarden-crds -n kubewarden
----

. Install the unified Helm chart:
+
[source,sh]
----
helm install kubewarden kubewarden/admission-controller -n kubewarden
----
+
The new single Helm chart also merges the three values files into one, and your
old values still apply. To get the same stack back, combine the values from the

> old three charts into one file and pass it to the install:

Reviewer: Only the `kubewarden-controller` and the `kubewarden-defaults` charts had values, isn't it?

I would show something like

```console
cat kubewarden-controller-values.yml kubewarden-default-values.yaml > admission-controller.yml
```

And then reference this file on the next `helm install` snippet.

Author: No, `kubewarden-crds` has flags to enable/disable CRDs.  https://github.com/kubewarden/helm-charts/blob/d8a12e9b96e992e3154c93642c050c257b31be73/charts/kubewarden-crds/values.yaml#L1-L12

+
[source,sh]
----
helm install kubewarden kubewarden/admission-controller -n kubewarden --values all-values.yaml
----

. Restore policies and policy servers:
+
[source,sh]
----
kubectl apply -f policyservers-backup.yaml
kubectl apply -f clusteradmissionpolicies-backup.yaml
kubectl apply -f admissionpolicies-backup.yaml
kubectl apply -f clusteradmissionpolicygroups-backup.yaml
kubectl apply -f admissionpolicygroups-backup.yaml
----

Once the controller reconciles them, the stack is back to where it was.

> == Migration script

Reviewer: We're not explaining to the user how to get the migration script. How are going to deliver it?

Right now the script is inside of the helm chart itself. I think it's a good idea, I would just move it from the root of the chart to a folder like `scripts`.

The we can tell the user to do something like that:

```console
helm pull my-repo/my-chart --version 2.0.0 --untar
cd my-chart
./scripts/migrate.sh
```

This instructions are just something I got from AI. They do make sense though

Author: Fine by me. I'll wait to merge the PRs updating the migration script before moving it and updating the docs.

The {admc-short-name} team ships a script that runs the whole migration to the unified
chart with no downtime. Check the prerequisites below before you run it.

If you can't meet one of the prerequisites, follow the "manual reinstallation"

> instructions for your own safety

Reviewer: ```suggestion
instructions for your own safety.
```

[WARNING]
====
The {admc-short-name} team test the migration path as much as possible. But it is
still recommended to backup policies and policy server definitions just in
case something unexpected happens. Therefore, it will be easier to restore to
the previous state if necessary.
====

The migration runs a few Helm operations in order: it annotates the stored
release manifests so the resources survive, uninstalls the legacy releases
without running their cleanup hooks, then installs the unified chart so it adopts
what is left. Order matters here. The annotations have to be in the stored
manifest, not just on the live objects, or Helm ignores them, and the hooks have

> to be skipped or they delete your {policy-server} resources.

Reviewer: ```suggestion
what is left.
```

This page is pretty long, IMHO we have to avoid all the unnecessary things to ensure users are actually going to read it.

The unified chart installs as a fresh release named `admission-controller`. Its
stateless plumbing (Deployment, Services, ConfigMap, RBAC, webhook configs) is
recreated under the chart's own names. The data and identity resources keep their
fixed names and are adopted in place: the CRDs, the `kubewarden-ca` Secret, the
`policy-server` RBAC, and the default {policy-server} with its recommended policies.
Because the plumbing is recreated rather than adopted, there is no immutable
selector to clash on and no `nameOverride` to carry around forever. A wrong step
here can delete resources or break admission, so the script does every step and

> checks it.

Reviewer: I would also remove that.

Unrelated, is something to be done when the old deployments have been done with `nameOverride` or `fullnameOverride`?

Author: The fields are dropped from the old helm chart values. The values fields used in the installation do not have them... Now that I'm thinking about that, as now the helm chart remove these fields from the selector, maybe we can add them again in the values with no big issues. I'll validate that.

Reviewer 2: I agree with removing info about the selectors and so, feels too low level.

=== What survives the migration

The following resources are adopted in place, with the same UIDs:

* The five {admc-short-name} CRDs.
* Your custom {policy-server} instances and policy CRs, along with their
  `Validating`/`MutatingWebhookConfiguration` objects.
* The `default` {policy-server} and recommended policy CRs (if enabled). The
  unified chart's DefaultsApplier adopts them in place. The
  `policy-server-default` Deployment is owned by the {policy-server} CR through
  an owner reference, so as long as the CR survives, the Deployment stays up

>   and admission continues.

Reviewer: ```suggestion
* If enabled, the `default` {policy-server} and the recommended policies. The
  deployment of the `default` {policy-server} is not going to be restarted by
  the migration, since no change is done the `PolicyServer` that defines it.
```

* The `kubewarden-ca` Secret. The unified chart's webhook template looks it up
  by name and reuses the existing CA data, so the policy-server pods that are

>   already running stay trusted by the webhook CA bundles. The CA is not rotated.

Reviewer: I think we can omit that.

=== What is recreated

The following resources are not adopted and are recreated:

* The controller's stateless plumbing, meaning its Deployment, Services, and
  ConfigMap, is deleted with the legacy release and recreated by the unified
  install under the chart's own names. The controller restarts briefly, but
  policy enforcement on workloads keeps working because the running policy-server
  pods serve it.
* The controller's leaf certs `kubewarden-webhook-server-cert` and
  `kubewarden-audit-scanner-client-cert` are regenerated from the preserved CA.
  They are not kept because the webhook server cert's SAN is tied to the webhook
  Service DNS name, and that name changes when the plumbing is recreated under
  the chart's own names. A stale cert would only be valid for the old Service
  name and would break the controller's admission webhooks. Since the CA itself
  does not change, the regenerated certs keep the webhook caBundle valid and

>   trust is not disrupted.

Reviewer: I would keep it super short

```suggestion
The deployment running the actual controller and some of it ancillary
resources (Service, ConfigMap, Secret) are recreated by the migration
script.

As a result of that, during the upgrade phase no new policy nor {policy-server}
can be created.
Existing {policy-server} instances and policies continue to operate as usual,
but cannot be edited.
```

=== Prerequisites

* Helm v4+ (needed for Server-Side Apply and post-renderer plugins)
* `kubectl` with access to your cluster
* `yq` v4 (https://github.com/mikefarah/yq[github.com/mikefarah/yq]) for the post-renderer
* `jq` for detecting installed chart versions

> * The three legacy releases must be installed in your cluster

Reviewer: ```suggestion
* The three legacy helm chart releases must be installed in your cluster
```

=== Script phases

The script runs five phases:

. *Preflight*: checks that the required tools are available, connects to the
  cluster, finds the three legacy releases, and warns if they are not at the
  latest chart version.
. *Annotation injection*: creates a Helm 4 post-renderer plugin that wraps
  `yq`, then runs `helm upgrade --reuse-values --post-renderer` on each legacy
  release to write `helm.sh/resource-policy: keep` into the stored release
  manifests. For the `kubewarden-crds` and `kubewarden-defaults` releases this
  keeps *all* rendered resources; for the `kubewarden-controller` release it
  keeps *only the `kubewarden-ca` Secret*, so the controller's stateless
  plumbing and its leaf certs are deleted (and later recreated/regenerated)

>   rather than adopted.

Reviewer: This feels too low level IMHO, mostly for a reviewer of the script (which is already well documented AFAIK).

```suggestion
  release to ensure a specific selection of resources they created
  survive the uninstall phase.
```

. *Legacy uninstall*: runs `helm uninstall --no-hooks` in reverse order. The
  `--no-hooks` flag skips the controller chart's pre-delete hook, which would
  otherwise delete all {policy-server} resources. Kept resources stay live; the

>   controller plumbing is removed.

Reviewer: ```suggestion
  otherwise delete all {policy-server} resources. Kept resources stay live. The
  controller and its ancillary resources are removed.
```

. *Unified chart install*: runs `helm install --take-ownership` as a fresh
  release named `admission-controller`, with no name/fullname overrides. Helm
  4's Server-Side Apply adopts the kept data/identity resources, updates their
  ownership metadata, and removes the legacy keep annotations; the controller
  plumbing is created fresh under the chart's natural names.
. *Verification*: checks that CRDs are owned by the new release, RBAC and the
  `kubewarden-ca` Secret were adopted in place (UIDs did not change), the
  controller Deployment is running (located via its hardened label selector),
  and the DefaultsApplier has labeled the default {policy-server} and
  recommended policies.

> 

Reviewer: I would add a small paragraph after the bullet point list to remind that, if something goes wrong, the user can recover by removing the helm chart release and then follow the manual upgrade instructions.

  --unified-chart kubewarden/admission-controller --interactive
----

Passing custom values to the unified chart:

[source,sh]
----
./kubewarden-unified-adm-controller-chart-migration.sh \
  --unified-chart kubewarden/admission-controller \
  --set "image.tag=v2.0.0" \
  --values ./my-custom-values.yaml
----

==== Available flags

[cols="1,2"]
|===
| Flag | Description

| `--unified-chart PATH_OR_NAME`
| Required. Local tarball or Helm repo chart name

| `--namespace NS`
| Namespace of the {admc-short-name} installation (default: `kubewarden`)

| `--kube-context CTX`
| Kubernetes context to use (default: current context)

| `--repo-name NAME`
| Helm repo name (default: `kubewarden`)

| `--repo-url URL`
| Helm repo URL (default: `https://charts.kubewarden.io`)

| `--timeout DURATION`
| Timeout for Helm operations (default: `5m`)

| `--set KEY=VALUE`
| Set a value for the unified chart install (repeatable)

| `--values FILE` / `-f FILE`
| Values file for the unified chart install (repeatable)

| `--interactive`
| Pause for confirmation before destructive steps

| `--dry-run`
| Show what would be done without making changes

| `--help`
| Print usage information

|===

> == Caveats

Reviewer: I would move this section about the limitations right at top, before the prerequisites. The user must be aware of these limitations ASAP

> *Plain release, no overrides.* The migrated release stores no `nameOverride` or

Reviewer: This is just a matter of personal taste, I don't know what @jhkrug thinks about that...

AI generated code has this habit of doing long sections and use bold formatting to create breaks.
In this case, I would just have each paragraph become a `h3`. This would also allow to split the really long paragraphs into smaller ones and, when needed, add bullet points.

Author: I've liked the suggestion and did the change. That also give a nicer navigation menu in the page. ;)

Reviewer 3: Yes, it's better semantic structure to use the h3.  And now, when the AI reads it, the structure is explicit and does not need to be inferred (possibly incorrectly) from style formatting.

include::partial$variables.adoc[]

= Migration from three-chart setup
:page-languages: [en, de, es, fr, ja, pt, zh]
:revdate: 2026-06-26
:page-revdate: {revdate}
:description: Migrate from the legacy kubewarden-crds, kubewarden-controller, and kubewarden-defaults Helm charts to the unified admission-controller chart.
:doc-persona: ["kubewarden-operator"]
:doc-topic: ["chart-migration"]
:doc-type: ["howto"]
:keywords: ["{admc-name}", "kubewarden", "helm", "chart", "migration", "upgrade", "kubewarden-crds", "kubewarden-controller", "kubewarden-defaults", "admission-controller"]
:sidebar_label: Migration from three-chart setup
:sidebar_position: 10
:current-version: {page-origin-branch}

Starting from Kubewarden 1.37, a single helm chart is used to manage
the stack.

If you are running the legacy three-chart setup (`kubewarden-crds`,

> `kubewarden-controller`, `kubewarden-defaults`), you have upgrade paths:

Reviewer: ```suggestion
`kubewarden-controller`, `kubewarden-defaults`), you have two possible upgrade paths:
```

* Use the migration script (`kubewarden-unified-adm-controller-chart-migration.sh`):
  this performs the migration without any downtime.
* Follow the reinstallation instructions: cleaner approach, but involves a short
  time during which your cluster won't be protected by policies.

[#manual-reinstallation]
== Manual reinstallation

This is the simplest approach. It consists of:

1. Doing a backup of your policies
2. Uninstall the old helm charts
3. Install the new single helm chart
4. Restore all your policies

> During step 2 and 4 your cluster won't be protected by any policy.

Reviewer: ```suggestion
After step #2 your cluster won't be protected. It will be
protected again once all your policies are restored.

You can then rely on Audit Scanner to find the non-compliant
resources that might have slipped into the cluster during this
short window.
```

+
[source,sh]
----
FILTER='del(.items[].metadata.uid, .items[].metadata.resourceVersion, .items[].metadata.creationTimestamp, .items[].metadata.generation, .items[].metadata.managedFields, .items[].status)'

kubectl get clusteradmissionpolicies -A -o yaml | yq "$FILTER" > clusteradmissionpolicies-backup.yaml
kubectl get admissionpolicies -A -o yaml | yq "$FILTER" > admissionpolicies-backup.yaml
kubectl get clusteradmissionpolicygroups -A -o yaml | yq "$FILTER" > clusteradmissionpolicygroups-backup.yaml
kubectl get admissionpolicygroups -A -o yaml | yq "$FILTER" > admissionpolicygroups-backup.yaml
kubectl get policyservers -A -o yaml | yq "$FILTER" > policyservers-backup.yaml
----

. Uninstall the old Helm charts:
+
[source,sh]
----
helm uninstall kubewarden-defaults -n kubewarden
helm uninstall kubewarden-controller -n kubewarden
helm uninstall kubewarden-crds -n kubewarden
----

. Install the unified Helm chart:
+
[source,sh]
----
helm install kubewarden kubewarden/admission-controller -n kubewarden
----
+
The new single Helm chart also merges the three values files into one, and your
old values still apply. To get the same stack back, combine the values from the

> old three charts into one file and pass it to the install:

Reviewer: ```suggestion
The new single Helm chart consolidates the three previous values files into one.
Your existing configuration settings remain valid; simply combine the values from
the three original files into a single file and provide it during installation:
```

+
[source,sh]
----
cat kubewarden-controller-values.yml kubewarden-default-values.yaml kubewarden-crds-values.yaml > admission-controller.yml
helm install kubewarden kubewarden/admission-controller -n kubewarden --values admission-controller.yaml
----

. Restore policies and policy servers:
+
[source,sh]
----
kubectl apply -f policyservers-backup.yaml
kubectl apply -f clusteradmissionpolicies-backup.yaml
kubectl apply -f admissionpolicies-backup.yaml
kubectl apply -f clusteradmissionpolicygroups-backup.yaml
kubectl apply -f admissionpolicygroups-backup.yaml
----

> Once the controller reconciles them, the stack is back to where it was.

Reviewer: ```suggestion
Now you just have to wait for the reconciliation of these resources to
be done. You're cluster has been successfully migrated.
```

== Migration script

The {admc-short-name} team ships a script that runs the whole migration to the unified
chart with no downtime. Check the prerequisites below before you run it.

If you can't meet one of the prerequisites, follow the "manual reinstallation"
instructions for your own safety.

[WARNING]
====
The {admc-short-name} team test the migration path as much as possible. But it is
still recommended to backup policies and policy server definitions just in
case something unexpected happens. Therefore, it will be easier to restore to

> the previous state if necessary.

Reviewer: ```suggestion
The {admc-short-name} team tested the migration path as much as possible.
However, we still recommended to backup policies and policy server definitions
before the migration takes place.
```

I've just noticed we don't provide instructions about how to backup resources.

IMHO we can either:
- point to the backup and restore instructions of the manual upgrade (fine with me)
- extend the migration script to perform the backup

Author: I'll update the docs to point to the instructions to the manual upgrade. If later we update the migration script, I'll update the docs as well.

====

The migration runs a few Helm operations in order: it annotates the stored
release manifests so the resources survive, uninstalls the legacy releases
without running their cleanup hooks, then installs the unified chart so it adopts

> what is left.

Reviewer: ```suggestion
what was left behind by the old charts.
```

`fullnameOverride`. The plumbing is recreated under the chart's own names instead
of being adopted in place, so there is no immutable selector to preserve. The
script builds the values file from `helm get values` on the three legacy releases
(`kubewarden-crds`, `kubewarden-controller`, `kubewarden-defaults`) and strips
any name or fullname overrides, so it carries only the configuration you set
yourself. Later upgrades need nothing special: a plain
`helm upgrade admission-controller ... --reuse-values` is enough. The release is
named `admission-controller`, and so are its controller resources.

==== Controller gap
Between the legacy uninstall and the new chart becoming
ready, no controller is running. The surviving policy-server pods still serve
the existing webhook configurations, so admission for active policies keeps
working. Any policy and policy server CR you create during this window will
return errors from the API server.

==== Settings drift
The DefaultsApplier rewrites each recommended policy's spec to
match the values you pass the unified chart. If you had changed `mode`,
`settings`, or other fields on those policies, pass the same values with `--set`

> or `--values` so your configuration survives.

Reviewer: I'm confused because the migration script keeps the values of the old charts, hence the user shouldn't have to bother about rewriting the `values.yml` or passing `--set` flags.

Are you referring to the case of someone
who deployed the recommended policies in `monitor` mode, but then changed their mode
to `protect` manually, like without using `helm upgrade`?

If that's the case, just remove the full section.

==== Renamed policies
The unified chart's default policy names match the legacy
defaults. If you renamed any through legacy values, pass the same `name`
overrides with `--set` or `--values`. Otherwise the applier removes the
old-named CRs after install.

==== Custom RBAC

> If you added extra permissions to `kubewarden-context-watcher` by

Reviewer: Again, I thought we pulled the old values from the helm release

  controller and its ancillary resources are removed.
. *Unified chart install*: runs `helm install --take-ownership` as a fresh
  release named `admission-controller`, with no name/fullname overrides. When
  the chart source is a Helm repo reference, the script resolves the chart
  version whose appVersion is `v1.37.0` and pins the install to that version.
  Helm 4's Server-Side Apply adopts the kept data/identity resources, updates
  their ownership metadata, and removes the legacy keep annotations; the
  controller plumbing is created fresh under the chart's natural names.
. *Verification*: checks that CRDs are owned by the new release, RBAC and the
  `kubewarden-ca` Secret were adopted in place (UIDs did not change), the
  controller Deployment is running (located via its hardened label selector),
  and the DefaultsApplier has labeled the default {policy-server} and
  recommended policies.

[WARNING]
====
If users face any issue during the migration using the script, they always can
restore the environment using the <<manual-reinstallation, `manual
reinstallation`>> instructions. Remember to backup your policies and policy

> server definition for such scenarios

Reviewer: ```suggestion
server definition for such scenarios.
```

:sidebar_label: Migration from three-chart setup
:sidebar_position: 10
:current-version: {page-origin-branch}

Starting from Kubewarden 1.37, a single helm chart is used to manage
the stack.

If you are running the legacy three-chart setup (`kubewarden-crds`,
`kubewarden-controller`, `kubewarden-defaults`), you have two possible upgrade paths:

* Use the migration script (`kubewarden-unified-adm-controller-chart-migration.sh`):
  this performs the migration without any downtime.
* Follow the reinstallation instructions: cleaner approach, but involves a short
  time during which your cluster won't be protected by policies.

[WARNING]
====
It's important to remember that, as described at
xref:/reference/upgrade-path.adoc[upgrade path documentation],  version jumping
is not allowed. You must migration from {admc-short-name} v1.36 to v1.37
====

[#manual-reinstallation]
== Manual reinstallation

This is the simplest approach. It consists of:

1. Doing a backup of your policies
2. Uninstall the old helm charts
3. Install the new single helm chart

> 4. Restore all your policies

Reviewer 3: ```suggestion
. Doing a backup of your policies
. Uninstall the old helm charts
. Install the new single helm chart
. Restore all your policies
```

After step #2 your cluster won't be protected. It will be
protected again once all your policies are restored.

You can then rely on Audit Scanner to find the non-compliant
resources that might have slipped into the cluster during this

> short window.

Reviewer 3: ```suggestion
You can then use Audit Scanner to find the non-compliant
resources that might have slipped into the cluster during this
short window.
```

kubectl get clusteradmissionpolicies -A -o yaml | yq "$FILTER" > clusteradmissionpolicies-backup.yaml
kubectl get admissionpolicies -A -o yaml | yq "$FILTER" > admissionpolicies-backup.yaml
kubectl get clusteradmissionpolicygroups -A -o yaml | yq "$FILTER" > clusteradmissionpolicygroups-backup.yaml
kubectl get admissionpolicygroups -A -o yaml | yq "$FILTER" > admissionpolicygroups-backup.yaml
kubectl get policyservers -A -o yaml | yq "$FILTER" > policyservers-backup.yaml
----

. Uninstall the old Helm charts:
+
[source,sh]
----
helm uninstall kubewarden-defaults -n kubewarden
helm uninstall kubewarden-controller -n kubewarden
helm uninstall kubewarden-crds -n kubewarden
----

. Install the unified Helm chart:
+
[source,sh]
----
helm install kubewarden kubewarden/admission-controller -n kubewarden
----
+
The new single Helm chart consolidates the three previous values files into one.
Your existing configuration settings remain valid; simply combine the values from
the three original files into a single file and provide it during installation:
+
[source,sh]
----

> cat kubewarden-controller-values.yml kubewarden-default-values.yaml kubewarden-crds-values.yaml > admission-controller.yml

Reviewer 2: if we are calling `-values.yaml` for all values, I would do the same with the `admission-controller.yaml` -> `admission-controller-values.yaml`.

Also note, in the cat it was `.yml`, in the helm install it is `.yaml`.

helm install kubewarden kubewarden/admission-controller -n kubewarden --values admission-controller.yaml
----

. Restore policies and policy servers:
+
[source,sh]
----
kubectl apply -f policyservers-backup.yaml
kubectl apply -f clusteradmissionpolicies-backup.yaml
kubectl apply -f admissionpolicies-backup.yaml
kubectl apply -f clusteradmissionpolicygroups-backup.yaml
kubectl apply -f admissionpolicygroups-backup.yaml
----

Now you just have to wait for the reconciliation of these resources to

> be done. You're cluster has been successfully migrated.

Reviewer 3: ```suggestion
finish. You're cluster has been successfully migrated.
```

== Migration script

The {admc-short-name} team ships a script that runs the whole migration to the unified
chart with no downtime. Check the prerequisites below before you run it.

If you can't meet one of the prerequisites, follow the "manual reinstallation"
instructions for your own safety.

[WARNING]
====
The {admc-short-name} team test the migration path as much as possible. But it is
still recommended to backup policies and policy server definitions just in
case something unexpected happens. Therefore, it will be easier to restore to
the previous state if necessary.

Follow the commands from the <<reinstallation-steps, `manual reinstallation section`>>
to backup your resources.

====

The migration runs a few Helm operations in order: it annotates the stored

> release manifests so the resources survive, uninstalls the legacy releases

Reviewer 2: ```suggestion
release manifests so the resources survive uninstallation, uninstalls the legacy releases
```

Just in case, to be clear.

without running their cleanup hooks, then installs the unified chart so it adopts
what was left behind by the old charts.

The unified chart installs as a fresh release named `admission-controller`. Its
stateless plumbing (Deployment, Services, ConfigMap, RBAC, webhook configs) is

> recreated under the chart's own names. The data and identity resources keep their

Reviewer 2: ```suggestion
recreated under the chart's own names. The stateful data and identity resources keep their
```

Maybe worth to say it that way, in contrast with the previous line of stateless plumbing?
I feel it rounds up the picture, panting better that nothing has been left over without talking about it.

fixed names and are adopted in place: the CRDs, the `kubewarden-ca` Secret, the
`policy-server` RBAC, and the default {policy-server} with its recommended policies.
Because the plumbing is recreated rather than adopted, there is no immutable
selector to clash on and no `nameOverride` to carry around forever. A wrong step
here can delete resources or break admission, so the script does every step and
checks it.

=== What survives the migration

The following resources are adopted in place, with the same UIDs:

* The five {admc-short-name} CRDs.
* Your custom {policy-server} instances and policy CRs, along with their
  `Validating`/`MutatingWebhookConfiguration` and all other ancillary objects.
* If enabled, the `default` {policy-server} and the recommended policies. The
  deployment of the `default` {policy-server} is not going to be restarted by
  the migration, since no change is done the `PolicyServer` that defines it.

=== What is recreated

The deployment running the actual controller and some of it ancillary
resources (Service, ConfigMap, Secret) are recreated by the migration

> script.

Reviewer 3: ```suggestion
The deployment running the actual controller and some of its ancillary
resources (Service, ConfigMap, Secret) are recreated by the migration
script.
```

can be created.
Existing {policy-server} instances and policies continue to operate as usual,
but cannot be edited.

=== Caveats

==== Plain release, no overrides
The migrated release stores no `nameOverride` or
`fullnameOverride`. The plumbing is recreated under the chart's own names instead
of being adopted in place, so there is no immutable selector to preserve. The
script builds the values file from `helm get values` on the three legacy releases
(`kubewarden-crds`, `kubewarden-controller`, `kubewarden-defaults`) and strips
any name or fullname overrides, so it carries only the configuration you set
yourself. Later upgrades need nothing special: a plain
`helm upgrade admission-controller ... --reuse-values` is enough. The release is
named `admission-controller`, and so are its controller resources.

==== Controller gap
Between the legacy uninstall and the new chart becoming
ready, no controller is running. The surviving policy-server pods still serve
the existing webhook configurations, so admission for active policies keeps
working. Any policy and policy server CR you create during this window will
return errors from the API server.

> ==== Renamed policies

Reviewer 2: Is this section needed?
If we already tell the users that we carry their values for the old charts, this feels like repeating ourselves.

The unified chart's default policy names match the legacy

> defaults. If you renamed any through legacy values, pass the same `name`

Reviewer 3: Is the word 'through' necessary here?
```suggestion
defaults. If you renamed any legacy values, pass the same `name`
```

overrides with `--set` or `--values`. Otherwise the applier removes the
old-named CRs after install.

==== Manual settings changes lost
If you changed resources configuration by other means that is not Helm upgrade
commands, you will lost your configuration. In such cases, you must ensure that
your customizations will be defined in the values passed to the migration

> script.

Reviewer 3: ```suggestion
If you changed resources configuration by other means that are not Helm upgrade
commands, you will lose your configuration. In such cases, you must ensure that
your customizations are defined in the values passed to the migration
script.
```

  controller and its ancillary resources are removed.
. *Unified chart install*: runs `helm install --take-ownership` as a fresh
  release named `admission-controller`, with no name/fullname overrides. When
  the chart source is a Helm repo reference, the script resolves the chart
  version whose appVersion is `v1.37.0` and pins the install to that version.
  Helm 4's Server-Side Apply adopts the kept data/identity resources, updates
  their ownership metadata, and removes the legacy keep annotations; the
  controller plumbing is created fresh under the chart's natural names.
. *Verification*: checks that CRDs are owned by the new release, RBAC and the
  `kubewarden-ca` Secret were adopted in place (UIDs did not change), the
  controller Deployment is running (located via its hardened label selector),
  and the DefaultsApplier has labeled the default {policy-server} and
  recommended policies.

[WARNING]
====
If users face any issue during the migration using the script, they always can
restore the environment using the <<manual-reinstallation, `manual
reinstallation`>> instructions. Remember to backup your policies and policy

> server definition for such scenarios.

Reviewer 3: ```suggestion
If users face any issue during the migration using the script, they can
restore the environment using the <<manual-reinstallation, `manual
reinstallation`>> instructions. Remember to backup your policies and policy
server definition for such scenarios.
```
